ath79: add support for Devolo Magic 2 WIFI

mgiganto opened the pull request:
This path support Devolo Magic 2 WIFI, board devolo_dlan2-2400-ac.
This device is a plc wifi router/extender with 2 Ethernet ports,
has a G.hn PLC and uses LCMP protocol from Home Grid Forum.

Hardware:
   SoC:         AR9344
   CPU:         560 MHz
   Flash:       16 MiB (W25Q128JVSIQ)
   RAM:         128 MiB DDR2
   Ethernet:    2xLAN 10/100/1000
   PLC:         88LX5152 (MaxLinear G.hn)
   PLC Flash:   W25Q32JVSSIQ
   PLC Uplink:  1Gbps MIMO
   PLC Link:    RGMII 1Gbps (WAN)
   WiFi:        Atheros AR9340 2.4GHz 802.11bgn
                Atheros AR9882-BR4A 5GHz 802.11ac
   Switch:      QCA8337, Port0:CPU, Port2:PLC, Port3:LAN1, Port4:LAN2
   Button:      3x Buttons (Reset, wifi and plc)
   LED:         3x Leds (wifi, plc white, plc red)
   GPIO Switch: 11-PLC Pairing (Active Low)
                13-PLC Enable
                21-WLAN power

Recommendations: Configure and link your PLC with OEM firmware
BEFORE you flash the device. PLC configuration/link should
remain in a different memory and should work straight forward
after flashing.

Restrictions: PLC link detection to trigger plc red led is not
available. PLC G.hn chip is not compatible with open-plc-tools,
it uses LCMP protocol with AES-128 and requires a different
software.

Notes: Pairing should be possible with gpio switch. Default
configuration will trigger wifi led with 2.4Ghz wifi traffic
and plc white led with wan traffic.

Flash instruction (TFTP):
1. Set PC to fixed ip address 192.168.0.100
2. Download the sysupgrade image and rename it to uploadfile
3. Start a tftp server with the image file in its root directory
4. Turn off the router
5. Press and hold Reset button
6. Turn on router with the reset button pressed and wait ~15 seconds
7. Release the reset button and after a short time
   the firmware should be transferred from the tftp server
8. Allow 1-2 minutes for the first boot.

Signed-off-by: Manuel Giganto <mgigantoregistros@gmail.com>

mregistros committed:
ath79: add support for Devolo Magic 2 WIFI

This path support Devolo Magic 2 WIFI, board devolo_dlan2-2400-ac.
This device is a plc wifi router/extender with 2 Ethernet ports,
has a G.hn PLC and uses LCMP protocol from Home Grid Forum.

Hardware:
   SoC:         AR9344
   CPU:         560 MHz
   Flash:       16 MiB (W25Q128JVSIQ)
   RAM:         128 MiB DDR2
   Ethernet:    2xLAN 10/100/1000
   PLC:         88LX5152 (MaxLinear G.hn)
   PLC Flash:   W25Q32JVSSIQ
   PLC Uplink:  1Gbps MIMO
   PLC Link:    RGMII 1Gbps (WAN)
   WiFi:        Atheros AR9340 2.4GHz 802.11bgn
                Atheros AR9882-BR4A 5GHz 802.11ac
   Switch:      QCA8337, Port0:CPU, Port2:PLC, Port3:LAN1, Port4:LAN2
   Button:      3x Buttons (Reset, wifi and plc)
   LED:         3x Leds (wifi, plc white, plc red)
   GPIO Switch: 11-PLC Pairing (Active Low)
                13-PLC Enable
                21-WLAN power

Recommendations: Configure and link your PLC with OEM firmware
BEFORE you flash the device. PLC configuration/link should
remain in a different memory and should work straight forward
after flashing.

Restrictions: PLC link detection to trigger plc red led is not
available. PLC G.hn chip is not compatible with open-plc-tools,
it uses LCMP protocol with AES-128 and requires a different
software.

Notes: Pairing should be possible with gpio switch. Default
configuration will trigger wifi led with 2.4Ghz wifi traffic
and plc white led with wan traffic.

Flash instruction (TFTP):
1. Set PC to fixed ip address 192.168.0.100
2. Download the sysupgrade image and rename it to uploadfile
3. Start a tftp server with the image file in its root directory
4. Turn off the router
5. Press and hold Reset button
6. Turn on router with the reset button pressed and wait ~15 seconds
7. Release the reset button and after a short time
   the firmware should be transferred from the tftp server
8. Allow 1-2 minutes for the first boot.

Signed-off-by: Manuel Giganto <mgigantoregistros@gmail.com>

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/01_leds:
```diff
@@ -71,6 +71,11 @@ comfast,cf-e5)
+devolo,dlan2-2400-ac|\
+devolo,magic2wifi)
+	ucidef_set_led_netdev "plcw" "dLAN" "devolo:status:dlan" "eth0.1" "rx"
+	ucidef_set_led_wlan "wlan" "WLAN" "devolo:status:wlan" "phy1tpt"
```
This can be moved to the DTS.

mgiganto replied:
Is there a way to move this to DTS?
     ucidef_set_led_netdev "plcw" "dLAN" "devolo:status:dlan" "eth0.1" "rx"
This is what I have on the DTS: 
     linux,default-trigger = "netdev";
I don´t know how to specify eth0.1 and rx... (or ideally rx and tx)

adrianschmutzler replied:
I was talking about the phy1tpt, which you already have in DTS, so you can just remove that line.
For the netdev, I haven't seen a DTS version so far.

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -124,6 +124,11 @@ ath79_setup_interfaces()
 		;;
+	devolo,dlan2-2400-ac|\
+	devolo,magic2wifi)
+		ucidef_set_interface_lan "eth0"
```
This is not required, as you define "lan" in the switch statement below.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,214 @@
+
+/ {
+	model = "Devolo Magic 2 Wifi";
+	compatible = "devolo,dlan2-2400-ac", "devolo,magic2wifi", "qca,ar9344";
```
You have to decide here: You can only have _one_ name for device definition.
I would go for the actual device name here (devolo,magic-2-wifi): This would then be your _only_ value for the cases in the files above, for the compatible (despite soc), for the Makefile block name, and for the DTS name (with _ instead of ,).

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,214 @@
+		wlan {
+			label = "devolo:status:wlan";
+			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy1tpt";
```
Ah, the trigger is already there. So you can just remove the corresponding line in 01_leds.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,214 @@
+		wlan {
+			label = "devolo:status:wlan";
+			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy1tpt";
```
What about phy0tpt?

mgiganto replied:
There are only 3 leds, and two are together. I think it´s more important to use the second one for the uplink, and there are more devices using 2.4G than 5G...

adrianschmutzler replied:
Okay, so the vendor indicated both 2.4 GHz and 5 GHz with this LED?
Maybe someone else knows whether we already have an example case for that.

Note that the policy in OpenWrt is to keep the LEDs as the vendor uses them. So, even if you wanted to repurpose another LED for separating WiFi indication, it wouldn't be accepted anyway ...

mgiganto replied:
The vendor keeps the wifi led on or off all the time, user chooses the behaviour, and the same goes for the PLC led, but with PLC led vendor switch it to red if you lose the PLC uplink.
Since PLC uplink can´t be monitored yet, the closest thing to identify your uplink is watching for traffic on port 2, but it didn´t work when I tried, so I am doing it at the eth0...

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,214 @@
+			label = "devolo:status:dlan";
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "netdev";
+//			default-state = "on";
```
Please remove any comments in your code ...

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,214 @@
+		};
+		wifi_button {
+			label = "WIFI button";
+			//                        linux,code = <KEY_RFKILL>;
```
KEY_RFKILL sound more logical to me?
Anyway, decide for the correct version and remove the other then.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,214 @@
+			debounce-interval = <60>;
+		};
+	};
+
```
Remove empty line.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,214 @@
+			};
+		};
+	};
+
```
Remove empty line.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,214 @@
+
+};
+
+
```
Remove second empty line.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,214 @@
+	status = "okay";
+};
+
+
```
Remove second empty line.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,214 @@
+		device = <&gmac>;
+		switch-phy-swap = <1>;
+	};
+*/
```
Remove commented code.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,214 @@
+	status = "okay";
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&art 0x1002>;
+//	mtd-mac-address-increment = <(-2)>;
```
Without the comment, `mtd-mac-address = <&art 0x1002>;` should be obsolete, as this is the address read by cal-data anyway. Can you remove both mac-address lines and check whether the address is still correct.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,214 @@
+};
+
+
+
```
Remove empty lines at the end.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,214 @@
+	};
+};
+
+
```
Remove empty line.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -297,6 +297,17 @@ define Device/comfast_cf-wr650ac-v2
 endef
 TARGET_DEVICES += comfast_cf-wr650ac-v2
 
+define Device/devolo_dlan2-2400-ac
```
Adjust this based on the "new" compatible.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -297,6 +297,17 @@ define Device/comfast_cf-wr650ac-v2
+  DEVICE_VENDOR := Devolo
+  DEVICE_MODEL := Magic 2 WIFI
+  DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca988x-ct kmod-ledtrig-gpio kmod-gpio-button-hotplug kmod-leds-gpio gpiod-tools \
+                     ebtables kmod-ebtables kmod-ebtables-ipv4 kmod-ebtables-ipv6 hostapd-utils wifitoggle wpad ip-full
```
Please check which of those you really need ...

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -297,6 +297,17 @@ define Device/comfast_cf-wr650ac-v2
+  DEVICE_MODEL := Magic 2 WIFI
+  DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca988x-ct kmod-ledtrig-gpio kmod-gpio-button-hotplug kmod-leds-gpio gpiod-tools \
+                     ebtables kmod-ebtables kmod-ebtables-ipv4 kmod-ebtables-ipv6 hostapd-utils wifitoggle wpad ip-full
+  SUPPORTED_DEVICES += dlan2-2400-ac
```
This is only required if the device was already supported in ar71xx.

adrianschmutzler commented:
Hi,

please have a look at my comments above.
Despite, you have some typos in the first line of your commit message.

You seem to be unsure about the correct MAC address assignment.
1. Can you provide the MAC address assignment (LAN, WAN, WiFi) from stock firmware?
2. Can you do a hexdump of your art partition and report the MAC addresses (with location) you find there?

mregistros committed:
ath79: add support for Devolo Magic 2 WIFI

This patch support Devolo Magic 2 WIFI, board devolo_dlan2-2400-ac.
This device is a plc wifi router/extender with 2 Ethernet ports,
has a G.hn PLC and uses LCMP protocol from Home Grid Forum.

Hardware:
   SoC:         AR9344
   CPU:         560 MHz
   Flash:       16 MiB (W25Q128JVSIQ)
   RAM:         128 MiB DDR2
   Ethernet:    2xLAN 10/100/1000
   PLC:         88LX5152 (MaxLinear G.hn)
   PLC Flash:   W25Q32JVSSIQ
   PLC Uplink:  1Gbps MIMO
   PLC Link:    RGMII 1Gbps (WAN)
   WiFi:        Atheros AR9340 2.4GHz 802.11bgn
                Atheros AR9882-BR4A 5GHz 802.11ac
   Switch:      QCA8337, Port0:CPU, Port2:PLC, Port3:LAN1, Port4:LAN2
   Button:      3x Buttons (Reset, wifi and plc)
   LED:         3x Leds (wifi, plc white, plc red)
   GPIO Switch: 11-PLC Pairing (Active Low)
                13-PLC Enable
                21-WLAN power

MACs Details verified with the stock firmware:
   Radio1: 2.4 GHz &wmac     *:4c Art location: 0x1002
   Radio0: 5.0 GHz &pcie     *:4d Art location: 0x5006
   Ethernet        &ethernet *:4e = 2.4 GHz + 2
   PLC uplink      ---       *:4f = 2.4 GHz + 3
Label MAC address is from PLC uplink

OEM SSID: echo devolo-$(grep SerialNumber /dev/mtd1 | grep -o ...$)
OEM WiFi password: grep DlanSecurityID /dev/mtd1|tr -d -|cut -d'=' -f 2

Recommendations: Configure and link your PLC with OEM firmware
BEFORE you flash the device. PLC configuration/link should
remain in different memory and should work straight forward
after flashing.

Restrictions: PLC link detection to trigger plc red led is not
available. PLC G.hn chip is not compatible with open-plc-tools,
it uses LCMP protocol with AES-128 and requires different
software.

Notes: Pairing should be possible with gpio switch. Default
configuration will trigger wifi led with 2.4Ghz wifi traffic
and plc white led with wan traffic.

Flash instruction (TFTP):
 1. Set PC to fixed ip address 192.168.0.100
 2. Download the sysupgrade image and rename it to uploadfile
 3. Start a tftp server with the image file in its root directory
 4. Turn off the router
 5. Press and hold Reset button
 6. Turn on router with the reset button pressed and wait ~15 seconds
 7. Release the reset button and after a short time
    the firmware should be transferred from the tftp server
 8. Allow 1-2 minutes for the first boot.

Signed-off-by: Manuel Giganto <mgigantoregistros@gmail.com>

mregistros committed:
ath79: add support for Devolo Magic 2 WIFI

This patch support Devolo Magic 2 WIFI, board devolo_dlan2-2400-ac.
This device is a plc wifi router/extender with 2 Ethernet ports,
has a G.hn PLC and uses LCMP protocol from Home Grid Forum.

Hardware:
   SoC:         AR9344
   CPU:         560 MHz
   Flash:       16 MiB (W25Q128JVSIQ)
   RAM:         128 MiB DDR2
   Ethernet:    2xLAN 10/100/1000
   PLC:         88LX5152 (MaxLinear G.hn)
   PLC Flash:   W25Q32JVSSIQ
   PLC Uplink:  1Gbps MIMO
   PLC Link:    RGMII 1Gbps (WAN)
   WiFi:        Atheros AR9340 2.4GHz 802.11bgn
                Atheros AR9882-BR4A 5GHz 802.11ac
   Switch:      QCA8337, Port0:CPU, Port2:PLC, Port3:LAN1, Port4:LAN2
   Button:      3x Buttons (Reset, wifi and plc)
   LED:         3x Leds (wifi, plc white, plc red)
   GPIO Switch: 11-PLC Pairing (Active Low)
                13-PLC Enable
                21-WLAN power

Recommendations: Configure and link your PLC with OEM firmware
BEFORE you flash the device. PLC configuration/link should
remain in different memory and should work straight forward
after flashing.

Restrictions: PLC link detection to trigger plc red led is not
available. PLC G.hn chip is not compatible with open-plc-tools,
it uses LCMP protocol with AES-128 and requires different
software.

Notes: Pairing should be possible with gpio switch. Default
configuration will trigger wifi led with 2.4Ghz wifi traffic
and plc white led with wan traffic.

Flash instruction (TFTP):
1. Set PC to fixed ip address 192.168.0.100
2. Download the sysupgrade image and rename it to uploadfile
3. Start a tftp server with the image file in its root directory
4. Turn off the router
5. Press and hold Reset button
6. Turn on router with the reset button pressed and wait ~15 seconds
7. Release the reset button and after a short time
   the firmware should be transferred from the tftp server
8. Allow 1-2 minutes for the first boot.

Signed-off-by: Manuel Giganto <mgigantoregistros@gmail.com>

mregistros committed:
ath79: add support for Devolo Magic 2 WIFI

This patch support Devolo Magic 2 WIFI, board devolo_dlan2-2400-ac.
This device is a plc wifi router/extender with 2 Ethernet ports,
has a G.hn PLC and uses LCMP protocol from Home Grid Forum.

Hardware:
   SoC:         AR9344
   CPU:         560 MHz
   Flash:       16 MiB (W25Q128JVSIQ)
   RAM:         128 MiB DDR2
   Ethernet:    2xLAN 10/100/1000
   PLC:         88LX5152 (MaxLinear G.hn)
   PLC Flash:   W25Q32JVSSIQ
   PLC Uplink:  1Gbps MIMO
   PLC Link:    RGMII 1Gbps (WAN)
   WiFi:        Atheros AR9340 2.4GHz 802.11bgn
                Atheros AR9882-BR4A 5GHz 802.11ac
   Switch:      QCA8337, Port0:CPU, Port2:PLC, Port3:LAN1, Port4:LAN2
   Button:      3x Buttons (Reset, wifi and plc)
   LED:         3x Leds (wifi, plc white, plc red)
   GPIO Switch: 11-PLC Pairing (Active Low)
                13-PLC Enable
                21-WLAN power

Recommendations: Configure and link your PLC with OEM firmware
BEFORE you flash the device. PLC configuration/link should
remain in different memory and should work straight forward
after flashing.

Restrictions: PLC link detection to trigger plc red led is not
available. PLC G.hn chip is not compatible with open-plc-tools,
it uses LCMP protocol with AES-128 and requires different
software.

Notes: Pairing should be possible with gpio switch. Default
configuration will trigger wifi led with 2.4Ghz wifi traffic
and plc white led with wan traffic.

Flash instruction (TFTP):
1. Set PC to fixed ip address 192.168.0.100
2. Download the sysupgrade image and rename it to uploadfile
3. Start a tftp server with the image file in its root directory
4. Turn off the router
5. Press and hold Reset button
6. Turn on router with the reset button pressed and wait ~15 seconds
7. Release the reset button and after a short time
   the firmware should be transferred from the tftp server
8. Allow 1-2 minutes for the first boot.

Signed-off-by: Manuel Giganto <mgigantoregistros@gmail.com>

mregistros committed:
ath79: add support for Devolo Magic 2 WIFI

This patch support Devolo Magic 2 WIFI, board devolo_dlan2-2400-ac.
This device is a plc wifi router/extender with 2 Ethernet ports,
has a G.hn PLC and uses LCMP protocol from Home Grid Forum.

Hardware:
   SoC:         AR9344
   CPU:         560 MHz
   Flash:       16 MiB (W25Q128JVSIQ)
   RAM:         128 MiB DDR2
   Ethernet:    2xLAN 10/100/1000
   PLC:         88LX5152 (MaxLinear G.hn)
   PLC Flash:   W25Q32JVSSIQ
   PLC Uplink:  1Gbps MIMO
   PLC Link:    RGMII 1Gbps (WAN)
   WiFi:        Atheros AR9340 2.4GHz 802.11bgn
                Atheros AR9882-BR4A 5GHz 802.11ac
   Switch:      QCA8337, Port0:CPU, Port2:PLC, Port3:LAN1, Port4:LAN2
   Button:      3x Buttons (Reset, wifi and plc)
   LED:         3x Leds (wifi, plc white, plc red)
   GPIO Switch: 11-PLC Pairing (Active Low)
                13-PLC Enable
                21-WLAN power

Recommendations: Configure and link your PLC with OEM firmware
BEFORE you flash the device. PLC configuration/link should
remain in different memory and should work straight forward
after flashing.

Restrictions: PLC link detection to trigger plc red led is not
available. PLC G.hn chip is not compatible with open-plc-tools,
it uses LCMP protocol with AES-128 and requires different
software.

Notes: Pairing should be possible with gpio switch. Default
configuration will trigger wifi led with 2.4Ghz wifi traffic
and plc white led with wan traffic.

Flash instruction (TFTP):
1. Set PC to fixed ip address 192.168.0.100
2. Download the sysupgrade image and rename it to uploadfile
3. Start a tftp server with the image file in its root directory
4. Turn off the router
5. Press and hold Reset button
6. Turn on router with the reset button pressed and wait ~15 seconds
7. Release the reset button and after a short time
   the firmware should be transferred from the tftp server
8. Allow 1-2 minutes for the first boot.

Signed-off-by: Manuel Giganto <mgigantoregistros@gmail.com>

mgiganto commented:
I tested this weekend these changes and they work fine. 
I also verified again the macs and they do match with the original firmware.
@adrianschmutzler I do want to thank you for all of your comments, they were very clear and useful, I appreciate your time on them :)

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,194 @@
+&pcie {
+	status = "okay";
+
+	ath9k: wifi@0,0 {
```
Is it ath9k or ath10k? Above you are using ath10k caldata extraction.
Despite, you do not seem to use this label, so maybe just remove it and use "wifi@0,0 {" here.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,194 @@
+	ath9k: wifi@0,0 {
+		compatible = "pci168c,003c";
+		reg = <0x0000 0 0 0 0>;
+		mtd-mac-address = <&art 0x5006>;
```
This should be set automatically when reading the caldata in 11-ath10k-caldata

Please remove this line and check.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_devolo_dlan2-2400-ac.dts:
```diff
@@ -0,0 +1,194 @@
+
+/ {
+	model = "Devolo Magic 2 Wifi";
+	compatible = "devolo,dlan2-2400-ac", "qca,ar9344";
```
Okay, I see that you changed everything consistently to "devolo,dlan2-2400-ac".
However, as I tried to point out in the initial comment, if you call the device "Magic 2 Wifi", you should use the same "name" for the identifier. So, please change "devolo,dlan2-2400-ac" to "devolo,magic-2-wifi" everywhere.

adrianschmutzler commented:
Hi, I left some new work for you.
Despite the comments, please squash your commits so that you only have commit in this Pull Request.
("git rebase -i @~X" with X being the number of commit to go back; in the "menu" you can then choose to squash commit into the previous one. Note that the first commit (the one to be squashed into) still needs to be marked with "pick". After that, re-push with "git push -f <remote> devolo_dlan2-2400-ac"...)

adrianschmutzler commented:
And please provide the MAC address overview I asked for above :-)

mregistros committed:
Changed the device to magic-2-wifi and some improvements

Signed-off-by: Manuel Giganto <mgigantoregistros@gmail.com>

mregistros committed:
Define label_mac

mregistros committed:
Merge branch 'devolo_dlan2-2400-ac' of https://github.com/mgiganto/openwrt into devolo_dlan2-2400-ac
